plugins/renepay/pay.c

```diff
@@ -0,0 +1,1798 @@
+#include "config.h"
+#include <ccan/array_size/array_size.h>
+#include <ccan/cast/cast.h>
+#include <ccan/htable/htable_type.h>
+#include <ccan/tal/str/str.h>
+#include <common/bolt11.h>
+#include <common/bolt12_merkle.h>
+#include <common/gossmap.h>
+#include <common/json_param.h>
+#include <common/json_stream.h>
+#include <common/memleak.h>
+#include <common/pseudorand.h>
+#include <common/type_to_string.h>
+#include <errno.h>
+#include <plugins/renepay/debug.h>
+#include <plugins/renepay/pay.h>
+#include <plugins/renepay/pay_flow.h>
+#include <plugins/renepay/uncertainty_network.h>
+
+// TODO(eduardo): maybe there are too many debug_err and plugin_err and
+// plugin_log(...,LOG_BROKEN,...) that could be resolved with a command_fail
+
+#define INVALID_ID UINT64_MAX
+#define MAX(a,b) ((a)>(b)? (a) : (b))
+
+static struct pay_plugin the_pay_plugin;
+struct pay_plugin * const pay_plugin = &the_pay_plugin;
+
+static void timer_kick(struct renepay * renepay);
+static struct command_result *try_paying(struct command *cmd,
+					 struct renepay * renepay,
+					 bool first_time);
+
+// TODO(eduardo): maybe we don't need these
+static void background_timer_kick(void*p UNUSED);
+static void background_settimer(void);
+
+void amount_msat_accumulate_(struct amount_msat *dst,
+			     struct amount_msat src,
+			     const char *dstname,
+			     const char *srcname)
+{
+	if (amount_msat_add(dst, *dst, src))
+		return;
+	debug_err("Overflow adding %s (%s) into %s (%s)",
+		   srcname, type_to_string(tmpctx, struct amount_msat, &src),
+		   dstname, type_to_string(tmpctx, struct amount_msat, dst));
+}
+
+void amount_msat_reduce_(struct amount_msat *dst,
+			 struct amount_msat src,
+			 const char *dstname,
+			 const char *srcname)
+{
+	if (amount_msat_sub(dst, *dst, src))
+		return;
+	debug_err("Underflow subtracting %s (%s) from %s (%s)",
+		   srcname, type_to_string(tmpctx, struct amount_msat, &src),
+		   dstname, type_to_string(tmpctx, struct amount_msat, dst));
+}
+
+
+#if DEVELOPER
+static void memleak_mark(struct plugin *p, struct htable *memtable)
+{
+	/* TODO(eduardo): understand the purpose of memleak_scan_obj, why use it
+	 * instead of tal_free?
+	 * 1st problem: this is executed before the plugin can process the
+	 * shutdown notification,
+	 * 2nd problem: memleak_scan_obj does not propagate to children.
+	 * For the moment let's just (incorrectly) do tal_free here
+	 * */
+	pay_plugin->ctx = tal_free(pay_plugin->ctx);
```

Comment:
It finds leaks at runtime.  You do not want to do this: if you are allocating things and not keeping a pointer to them, it's probably a bug!

Reply:
I'll remove this line and uncomment the `memleak_scan_obj`.